_posts/2022-05-12-faster-rand.md

```diff
@@ -0,0 +1,1096 @@
+             {exsp,next}:     16.8 ns     37.2%
+       {splitmix64,next}:    331.1 ns    734.3%
+           unique_phash2:     21.1 ns     46.8%
+                procdict:     75.2 ns    166.7%
+        {mwc59,procdict}:     16.6 ns     36.8%
+```
+In this section no range capping is done.  The raw generator output is used.
+
+Here we have the `dummy` generator, which is an undocumented generator
+within the `rand` plug-in framework that only does a minimal state
+update and returns a constant.  It is used here to measure
+plug-in framework overhead.
+
+The plug-in framework overhead is measured to 25.5 ns that matches
+`exsp`&nbsp;-&nbsp;`{exsp,next}`&nbsp;=&nbsp;23.0 ns fairly well,
+which is the same algorithm within and without the plug-in framework,
+giving another measure of the framework overhead.
+
+`procdict` is the default algorithm `exsss` but makes the plug-in
+framework store the generator state in the process dictionary,
+which here costs 30 ns.
+
+`{mwc59,procdict}` stores the generator state in the process dictionary,
+which here costs 12.9 ns. The state term that is stored is much smaller
+than for the plug-in framework.  Compare to `procdict`
+in the previous paragraph.
+
+
+
+Summary
+-------
+
+The [new fast] generator's functions in the `rand` module
+fills a niche for speed over quality where the type-based
+[JIT optimizations] have elevated the performance.
+
+The combination of high speed and high quality can only
+be fulfilled with a [BIF implementation], but we hope that
+to be a combination we do not need to address...
+
+[Implementing a PRNG] is tricky business.
+
+Recent improvements in [`rand_SUITE:measure/1`]
+highlights what the precious CPU cycles are used for.
+
+
+
+[Use `rand` anyway]:        #use-rand-anyway
+[BIF implementation]:       #write-a-bif
+[Write a BIF]:              #write-a-bif
+[Write a NIF]:              #write-a-nif
+[new fast]:                 #write-a-simple-prng
+[Use the system time]:      #use-the-system-time
+[Hash a "unique" value]:    #hash-a-unique-value
+[Write a simple PRNG]:      #write-a-simple-prng
+
+[Quality]:                  #quality
+[Spectral score]:           #spectral-score
+[PRNG tests]:               #prng-tests
+[Storing the state]:        #storing-the-state
+
+[JIT optimizations]:        #jit-optimizations
+[Implementing a PRNG]:      #implementing-a-prng
+[`rand_SUITE:measure/1`]:   #rand_suitemeasure1
+[Measurement results]:      #measurement-results
+
+[Looking for a faster RNG]: https://erlangforums.com/t/looking-for-a-faster-rng/
+
+[potatosalad]: https://github.com/potatosalad/
+
+[latrules]: https://www.iro.umontreal.ca/~lecuyer/myftp/papers/latrules.ps
+
+[TestU01]: http://simul.iro.umontreal.ca/testu01/
+
+[PractRand]: http://pracrand.sourceforge.net/
+
+[type-based optimizations]: https://www.erlang.org/blog/type-based-optimizations-in-the-jit/
+
+[Sebastiano Vigna]: https://vigna.di.unimi.it/
+
+[CPRNG]: https://github.com/vigna/CPRNG/
```

Comment:
How odd that these are all shown at the bottom of the blog post... maybe if we collapse them?

```suggestion
[Looking for a faster RNG]: https://erlangforums.com/t/looking-for-a-faster-rng/
[potatosalad]: https://github.com/potatosalad/
[latrules]: https://www.iro.umontreal.ca/~lecuyer/myftp/papers/latrules.ps
[TestU01]: http://simul.iro.umontreal.ca/testu01/
[PractRand]: http://pracrand.sourceforge.net/
[type-based optimizations]: https://www.erlang.org/blog/type-based-optimizations-in-the-jit/
[Sebastiano Vigna]: https://vigna.di.unimi.it/
[CPRNG]: https://github.com/vigna/CPRNG/
```

Reply:
Are they?  Where?  I looked at the deploy preview and there they were not shown, as expected...

Reply:
I don't know what I looked at, but they are gone now. So no problem then.